File: client/lib/data/repositories/auth/operations/create_account.dart
import 'package:dartz/dartz.dart';
import 'package:likha/core/database/db_schema.dart';
import 'package:likha/core/errors/failures.dart';
import 'package:likha/core/sync/mutation_result.dart';
import 'package:likha/core/sync/sync_queue.dart';
import 'package:likha/core/utils/typedef.dart';
import 'package:likha/data/datasources/local/auth/auth_local_datasource.dart';
import 'package:likha/data/models/auth/user_model.dart';
import 'package:likha/domain/auth/entities/user.dart';
import 'package:sqflite_sqlcipher/sqflite.dart';
import 'package:uuid/uuid.dart';

ResultFuture<MutationResult<User>> createAccount(
  AuthLocalDataSource localDataSource,
  SyncQueue syncQueue,
  {
    required String username,
    required String fullName,
    required String role,
  }) async {
  try {
    final now = DateTime.now();
    final userId = const Uuid().v4();
    final queueEntryId = const Uuid().v4();

    final optimisticUser = UserModel(
      id: userId,
      username: username,
      fullName: fullName,
      role: role,
      accountStatus: 'pending_activation',
      isActive: false,
      activatedAt: null,
      createdAt: now,
      updatedAt: now,
      cachedAt: now,
      syncStatus: SyncStatus.pending,
    );

    final db = await localDataSource.localDatabase.database;
    await db.transaction((txn) async {
      final map = optimisticUser.toMap();
      map[CommonCols.cachedAt] = now.toIso8601String();
      map[CommonCols.syncStatus] = SyncStatus.pending.dbValue;
      await txn.insert(
        DbTables.users,
        map,
        conflictAlgorithm: ConflictAlgorithm.replace,
      );

      await syncQueue.enqueue(
        SyncQueueEntry(
          id: queueEntryId,
          entityType: SyncEntityType.adminUser,
          operation: SyncOperation.create,
          payload: optimisticUser.toPayload(),
          status: SyncStatus.pending,
          retryCount: 0,
          maxRetries: 5,
          createdAt: now,
        ),
        txn: txn,
      );
    });


    return Right(MutationResult(entity: optimisticUser, status: SyncStatus.pending));
  } catch (e) {
    return Left(ServerFailure(e.toString()));
  }
}

File: server/src/modules/auth/repository_operations/user/create_account.rs
use chrono::Utc;
use sea_orm::*;
use uuid::Uuid;

use ::entity::users;
use crate::utils::{AppError, AppResult};

pub async fn create_account(
    db: &DatabaseConnection,
    username: String,
    first_name: String,
    last_name: String,
    role: String,
    client_id: Option<Uuid>,
) -> AppResult<users::Model> {
    let user = users::ActiveModel {
        id: Set(client_id.unwrap_or_else(Uuid::new_v4)),
        username: Set(username),
        password_hash: Set(None),
        first_name: Set(first_name),
        last_name: Set(last_name),
        role: Set(role),
        account_status: Set("pending_activation".to_string()),
        activated_at: Set(None),
        created_at: Set(Utc::now().naive_utc()),
        updated_at: Set(Utc::now().naive_utc()),
        deleted_at: Set(None),
    };

    user.insert(db)
        .await
        .map_err(|e| AppError::InternalServerError(format!("Failed to create account: {}", e)))
}
